grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/Atom.java

```diff
@@ -151,24 +151,22 @@ public int baseResolutionPriority(){
     /**
      * @return set of potentially applicable rules - does shallow (fast) check for applicability
      */
-    private Set<Rule> getPotentialRules(){
-        OntologyConcept ontologyConcept = getOntologyConcept();
-        return ontologyConcept != null ?
-                ontologyConcept.subs().flatMap(OntologyConcept::getRulesOfConclusion).collect(Collectors.toSet()) :
-                ReasonerUtils.getRules(graph());
+    private Stream<Rule> getPotentialRules(){
+        return new RuleGraph(graph()).getRulesWithType(getOntologyConcept());
```

Comment:
[*] This could potentially be `RuleUtil. getRulesWithType(graph, getOntologyConcept)` . . . i.e no object allocation needed